src/audio/mixin_mixout/mixin_mixout.c

```diff
@@ -250,76 +254,98 @@ static void silence(struct audio_stream *stream, uint32_t start_frame,
+		struct sof_sink *mixout_sink;
 		struct mixout_data *mixout_data;
 		struct processing_module *mixout_mod;
 		uint32_t free_frames;
 
+		/* WORKAROUND: since mixin is always connected to mixout, we can safely assume
+		 * mixin sink interface is implemented via comp_buffer. This is, of course,
+		 * not the case for other modules.
+		 * TODO: find out a solution to reach mixout without knowledge of mixin
+		 * sof_sink implementation.
+		 */
+		stream = container_of(sinks[i], struct audio_stream, sink_api);
 		/* unused buffer between mixin and mixout */
-		unused_in_between_buf_c = container_of(output_buffers[i].data,
-						       struct comp_buffer, stream);
-		mixout = unused_in_between_buf_c->sink;
-		sink_id = IPC4_SRC_QUEUE_ID(buf_get_id(unused_in_between_buf_c));
-
-		active_mixouts[i] = mixout;
-		sinks_ids[i] = sink_id;
+		unused_in_between_buf = container_of(stream, struct comp_buffer, stream);
+		mixout = unused_in_between_buf->sink;
 
-		sink = list_first_item(&mixout->bsink_list, struct comp_buffer, source_list);
+		/* Skip non-active mixout like it is not connected so it does not
+		 * block other possibly connected mixouts. In addition, non-active
+		 * mixouts might have their sink buffer/interface not yet configured.
+		 */
+		if (mixout->state != COMP_STATE_ACTIVE) {
+			active_mixouts[i] = NULL;
+			continue;
+		}
 
 		mixout_mod = comp_get_drvdata(mixout);
+		active_mixouts[i] = mixout_mod;
+		mixout_sink = mixout_mod->sinks[0];
+
+		/* mixout might be created on another pipeline. Its sink stream params are usually
+		 * configured in .prepare(). It is possible that such .prepare() was not yet called
+		 * for mixout pipeline. Hence the check above if mixout state is active. However,
+		 * let's just in case check here if sink stream params are really configured as
+		 * proceeding with unconfigured sink will lead to hard to debug bugs.
+		 * Unconfigured stream params are filled with zeros.
+		 * TODO: introduce something like sink_is_configured() ?
+		 */
+		if (!mixout_sink || sink_get_channels(mixout_sink) == 0) {
+			comp_err(dev, "mixout sink not configured!");
+			return -EINVAL;
+		}
+
+		sinks_ids[i] = IPC4_SRC_QUEUE_ID(buf_get_id(unused_in_between_buf));
+
 		mixout_data = module_get_private_data(mixout_mod);
 		pending_frames = get_mixin_pending_frames(mixout_data, dev);
 		if (!pending_frames) {
 			comp_err(dev, "No source info");
 			return -EINVAL;
 		}
 
-		/* Normally this should never happen as we checked above
-		 * that mixout is in active state and so its sink buffer
-		 * should be already initialized in mixout .params().
+		/* In theory, though unlikely, mixout sink can be connected to some module on
+		 * another core. In this case free space in mixout sink buffer can suddenly increase
+		 * (data consumed on another core) after the buffer was already acquired. Let's only
+		 * access free space that was at the moment of acquiring the buffer.
 		 */
-		if (!sink->hw_params_configured) {
```

Comment:
this is replaced by the check in line 327 above, right?

Reply:
Yes. As sink interface does not provide anything like hw_params_configured, I just use sink_get_channels() > 0 check instead. If stream parameters are not yet configured they are all filled with zeros and so channel number is 0. For configured stream params, channel number is always > 0.
Maybe somethings like sink_is_configured() should be introduced for this purpose.